### PR TITLE
Link deletion

### PR DESCRIPTION
**Pull request by cpranav93**
Adds the function for link_deletion based on a deletion probability to be passed as argument to the function. The method uses upper triangular matrices to perform tensor based link deletion. 

To check, line 29: do we need triu_adj.values as they are just 1 and will always be 1. 

There are multiple files in this PR simply because cookie cutter and dgl_framework are yet to merged. It should simplify after the merge.

### CODE REVIEW

File: dgl_ptm/dgl_ptm/link_deletion.py [hunk 1]
```diff
@@ -1,7 +1,70 @@
-#!/usr/bin/env python
-# coding: utf-8
+import dgl
+import torch
 
-# link_deletion - Randomly deletes links between agents
+def link_deletion(agent_graph, del_prob: float):
+    '''
+        link_deletion - Randomly deletes links between agents upto a deletion probablity 'del_prob'
```
Comment (meiertgrootes) [inline dgl_ptm/dgl_ptm/link_deletion.py:6]: deletes links between agents with a deletion probability 'del_prob' by samoliing against a random uniform distribution.

**Review by meiertgrootes — COMMENTED** [review]
Very nice. We should probably move the utility functions into a separate submodule, but that is hard to do before merging.

Otherwise just a small comment on the description

File: dgl_ptm/dgl_ptm/link_deletion.py (hunk 1 above)
Comment (cpranav93) [reply dgl_ptm/dgl_ptm/link_deletion.py]: Changed in latest commit

File: dgl_ptm/dgl_ptm/link_deletion.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+def _sparse_upper_triangular(spm):
+    """
+    select the upper triangular matrix from a sparse matrix
+
+    Args:
+        spm: the sparse matrix (dgl.sparse.SparseMatrix)
+        
+    Return: dgl.sparse.SparseMatrix
+    """
+    mask = spm.row <= spm.col
```
Comment (vanlankveldthijs) [inline dgl_ptm/dgl_ptm/link_deletion.py:58]: Should this be replaced by "mask = spm.row < spm.col" to also disable any stray self-loops?

Comment (cpranav93) [reply dgl_ptm/dgl_ptm/link_deletion.py:58]: I think yes. @meiertgrootes ?
